stan/math/opencl/rev/to_arena.hpp

```diff
@@ -0,0 +1,34 @@
+#ifndef STAN_MATH_OPENCL_REV_TO_ARENA_HPP
+#define STAN_MATH_OPENCL_REV_TO_ARENA_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/is_kernel_expression.hpp>
+#include <stan/math/opencl/rev/arena_type.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Converts given argument into a type that either has any dynamic allocation on
+ * AD stack or schedules its destructor to be called when AD stack memory is
+ * recovered.
+ *
+ * This overload is used for `matrix_cl`. It is converted to a subclass that
+ * schedules its destructor to be called when the memory is recovered.
+ *
+ * @tparam T type of scalar
+ * @param a argument
+ * @return argument
+ */
+template <typename T>
+arena_t<matrix_cl<T>> to_arena(const matrix_cl<T>& a) {
+  return arena_t<matrix_cl<T>>(const_cast<cl::Buffer&>(a.buffer()), a.rows(),
+                               a.cols(), a.view());
+}
```

Comment:
The buffer in a matrix_cl is not const so I'm not following why we need a const_cast here. Why not just have a case that makes a copy of the buffer for actual constant `matrix_cl`'s?

Reply:
We are accepting a const reference here. So the buffer in it is const. (Accepting a forwarding reference would also accept rvalues without adding the const, but it would not help if the argument this function is called with is actually const).

> Why not just have a case that makes a copy of the buffer for actual constant matrix_cl's?

Because we don't need to and it is inefficient to copy the buffer.

Reply:
Are there times where we would modify this? I'm mostly just worried about UB if we ever write to this buffer and it's actually const (if I'm reading [this](https://stackoverflow.com/a/54504302/2269255) correctly)

Reply:
I don't think we are ever declaring `matrix_cl`s `const`. Also the buffer object (and all other OpenCL objects) only contains an identifier (imagine int or a pointer) of an object allocated by the OpenCL implementation. Constness of the identifier does not affect constness of the actual memory (it can instead be set by flags, such as `CL_MEM_READ_ONLY` ...) So I am pretty sure it would just work even if we did what C++ consideres UB.

Reply:
Why don't we just make the buffer `mutable`? Then it doesn't need to be const cast.